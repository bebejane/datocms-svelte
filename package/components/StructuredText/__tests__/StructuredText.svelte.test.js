import { describe, expect, it } from 'vitest';
import '@testing-library/jest-dom';
import { render } from '@testing-library/svelte';
import { isBlock, isHeading, isInlineItem, isItemLink, isSpan } from 'datocms-structured-text-utils';
import { StructuredText } from '../../..';
import { heading, structuredTextWithBlocksAndLinks } from './__fixtures__/structuredText';
import CustomSpan from './__fixtures__/CustomSpan.svelte';
import IncreasedLevelHeading from './__fixtures__/IncreasedLevelHeading.svelte';
import ItemLink from './__fixtures__/ItemLink.svelte';
import Block from './__fixtures__/Block.svelte';
import InlineItem from './__fixtures__/InlineItem.svelte';
describe('StructuredText', () => {
    describe('with no value', () => {
        it('renders null', () => {
            const { container } = render(StructuredText, { props: { data: null } });
            expect(container).toMatchSnapshot();
        });
    });
    describe('with a very simple pure dast (only the `value` of a structured text)', () => {
        describe('with default rules', () => {
            it('renders the document', () => {
                const { container } = render(StructuredText, { props: { data: heading.value } });
                expect(container).toMatchSnapshot();
            });
        });
        describe('with custom mark rules', () => {
            it('renders the document', () => {
                const { container } = render(StructuredText, {
                    props: { data: heading, components: [[isSpan, CustomSpan]] }
                });
                expect(container).toMatchSnapshot();
            });
        });
    });
    describe('with a very simple dast', () => {
        describe('with default rules', () => {
            it('renders the document', () => {
                const { container } = render(StructuredText, { props: { data: heading } });
                expect(container).toMatchSnapshot();
            });
        });
        describe('with custom mark rules', () => {
            it('renders the document', () => {
                const { container } = render(StructuredText, {
                    props: { data: heading, components: [[isSpan, CustomSpan]] }
                });
                expect(container).toMatchSnapshot();
            });
        });
    });
    describe('with a dast with no links nor blocks', () => {
        describe('with default rules', () => {
            it('renders the document', () => {
                const { container } = render(StructuredText, {
                    props: { data: heading }
                });
                expect(container).toMatchSnapshot();
            });
        });
        describe('with custom rules', () => {
            it('renders the document', () => {
                const { container } = render(StructuredText, {
                    props: { data: heading, components: [[isHeading, IncreasedLevelHeading]] }
                });
                expect(container).toMatchSnapshot();
            });
        });
    });
    describe('with a dast including links and blocks', () => {
        describe('with default rules', () => {
            it('renders the document', () => {
                const { container } = render(StructuredText, {
                    props: {
                        data: structuredTextWithBlocksAndLinks,
                        components: [
                            [isInlineItem, InlineItem],
                            [isItemLink, ItemLink],
                            [isBlock, Block]
                        ]
                    }
                });
                expect(container).toMatchSnapshot();
            });
        });
        describe('with missing component for item links', () => {
            it('raises an error', () => {
                expect(() => {
                    render(StructuredText, {
                        props: {
                            data: structuredTextWithBlocksAndLinks,
                            components: [
                                [isBlock, Block],
                                [isInlineItem, InlineItem]
                            ]
                        }
                    });
                }).toThrowErrorMatchingSnapshot();
            });
        });
        describe('with missing component for inline items', () => {
            it('raises an error', () => {
                expect(() => {
                    render(StructuredText, {
                        props: {
                            data: structuredTextWithBlocksAndLinks,
                            components: [
                                [isBlock, Block],
                                [isItemLink, ItemLink]
                            ]
                        }
                    });
                }).toThrowErrorMatchingSnapshot();
            });
        });
        describe('with missing links', () => {
            it('raises an error', () => {
                expect(() => {
                    render(StructuredText, {
                        props: {
                            data: { ...structuredTextWithBlocksAndLinks, links: [] },
                            components: [
                                [isBlock, Block],
                                [isInlineItem, InlineItem],
                                [isItemLink, ItemLink]
                            ]
                        }
                    });
                }).toThrowErrorMatchingSnapshot();
            });
        });
        describe('with missing component for blocks', () => {
            it('raises an error', () => {
                expect(() => {
                    render(StructuredText, {
                        props: {
                            data: structuredTextWithBlocksAndLinks,
                            components: [
                                [isInlineItem, InlineItem],
                                [isItemLink, ItemLink]
                            ]
                        }
                    });
                }).toThrowErrorMatchingSnapshot();
            });
        });
        describe('with missing blocks', () => {
            it('raises an error', () => {
                expect(() => {
                    render(StructuredText, {
                        props: {
                            data: { ...structuredTextWithBlocksAndLinks, blocks: [] },
                            components: [
                                [isBlock, Block],
                                [isInlineItem, InlineItem],
                                [isItemLink, ItemLink]
                            ]
                        }
                    });
                }).toThrowErrorMatchingSnapshot();
            });
        });
    });
});
