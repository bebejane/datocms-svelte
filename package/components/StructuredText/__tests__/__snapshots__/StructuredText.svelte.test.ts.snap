// Vitest Snapshot v1

exports[`StructuredText > with a dast including links and blocks > with default rules > renders the document 1`] = `
<body>
  <div>
    <h1>
      This is a
       
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      <mark>
        title
         
        
        <!--&lt;Lines&gt;-->
        
        <!--&lt;Span&gt;-->
      </mark>
      
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      <a
        href="/docs/how-to-code"
      >
        How to code
      </a>
      ;
      
      <!--&lt;InlineItem&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      <a
        href="/docs/how-to-code"
        target="_blank"
      >
        here!
         
        
        <!--&lt;Lines&gt;-->
        
        <!--&lt;Span&gt;-->
        
        
        
        <!--&lt;Node&gt;-->
        
        
      </a>
      
      <!--&lt;ItemLink&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
    </h1>
    
    <!--&lt;Heading&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    <figure>
      <blockquote>
        Foo bar.
      </blockquote>
       
      <figcaption>
        Mark Smith
      </figcaption>
    </figure>
    
    <!--&lt;Block&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    
    
    <!--&lt;Root&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    
    <!--&lt;StructuredText&gt;-->
  </div>
</body>
`;

exports[`StructuredText > with a dast including links and blocks > with missing blocks > raises an error 1`] = `"The Structured Text document contains a 'block' node, but cannot find a record with ID 456 inside data.blocks!"`;

exports[`StructuredText > with a dast including links and blocks > with missing component for blocks > raises an error 1`] = `"The Structured Text document contains a 'block' node, but no component for rendering is specified!"`;

exports[`StructuredText > with a dast including links and blocks > with missing component for inline items > raises an error 1`] = `"The Structured Text document contains an 'inlineItem' node, but no component for rendering is specified!"`;

exports[`StructuredText > with a dast including links and blocks > with missing component for item links > raises an error 1`] = `"The Structured Text document contains an 'itemLink' node, but no component for rendering is specified!"`;

exports[`StructuredText > with a dast including links and blocks > with missing links > raises an error 1`] = `"The Structured Text document contains an 'itemLink' node, but cannot find a record with ID 123 inside data.links!"`;

exports[`StructuredText > with a dast with no links nor blocks > with custom rules > renders the document 1`] = `
<body>
  <div>
    <h2>
      This
       
      <br />
       
      is a 
       
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      <strong>
        title
         
        
        <!--&lt;Lines&gt;-->
        
        <!--&lt;Span&gt;-->
      </strong>
      
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
    </h2>
    
    <!--&lt;IncreasedLevelHeading&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    
    
    <!--&lt;Root&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    
    <!--&lt;StructuredText&gt;-->
  </div>
</body>
`;

exports[`StructuredText > with a dast with no links nor blocks > with default rules > renders the document 1`] = `
<body>
  <div>
    <h1>
      This
       
      <br />
       
      is a 
       
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      <strong>
        title
         
        
        <!--&lt;Lines&gt;-->
        
        <!--&lt;Span&gt;-->
      </strong>
      
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
    </h1>
    
    <!--&lt;Heading&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    
    
    <!--&lt;Root&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    
    <!--&lt;StructuredText&gt;-->
  </div>
</body>
`;

exports[`StructuredText > with a very simple dast > with custom mark rules > renders the document 1`] = `
<body>
  <div>
    <h1>
      This
       
      <br />
       
      is a 
       
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;CustomSpan&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      <b>
        title
         
        
        <!--&lt;Lines&gt;-->
        
        <!--&lt;CustomSpan&gt;-->
      </b>
      
      
      <!--&lt;CustomSpan&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
    </h1>
    
    <!--&lt;Heading&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    
    
    <!--&lt;Root&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    
    <!--&lt;StructuredText&gt;-->
  </div>
</body>
`;

exports[`StructuredText > with a very simple dast > with default rules > renders the document 1`] = `
<body>
  <div>
    <h1>
      This
       
      <br />
       
      is a 
       
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      <strong>
        title
         
        
        <!--&lt;Lines&gt;-->
        
        <!--&lt;Span&gt;-->
      </strong>
      
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
    </h1>
    
    <!--&lt;Heading&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    
    
    <!--&lt;Root&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    
    <!--&lt;StructuredText&gt;-->
  </div>
</body>
`;

exports[`StructuredText > with a very simple pure dast (only the \`value\` of a structured text) > with custom mark rules > renders the document 1`] = `
<body>
  <div>
    <h1>
      This
       
      <br />
       
      is a 
       
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;CustomSpan&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      <b>
        title
         
        
        <!--&lt;Lines&gt;-->
        
        <!--&lt;CustomSpan&gt;-->
      </b>
      
      
      <!--&lt;CustomSpan&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
    </h1>
    
    <!--&lt;Heading&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    
    
    <!--&lt;Root&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    
    <!--&lt;StructuredText&gt;-->
  </div>
</body>
`;

exports[`StructuredText > with a very simple pure dast (only the \`value\` of a structured text) > with default rules > renders the document 1`] = `
<body>
  <div>
    <h1>
      This
       
      <br />
       
      is a 
       
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Lines&gt;-->
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      <strong>
        title
         
        
        <!--&lt;Lines&gt;-->
        
        <!--&lt;Span&gt;-->
      </strong>
      
      
      <!--&lt;Span&gt;-->
      
      
      
      <!--&lt;Node&gt;-->
      
      
    </h1>
    
    <!--&lt;Heading&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    
    
    <!--&lt;Root&gt;-->
    
    
    
    <!--&lt;Node&gt;-->
    
    <!--&lt;StructuredText&gt;-->
  </div>
</body>
`;

exports[`StructuredText > with no value > renders null 1`] = `
<body>
  <div>
    
    <!--&lt;StructuredText&gt;-->
  </div>
</body>
`;
